### ISSUE TITLE
WebSocketException because WebSocket is in an invalid state

### ISSUE THREAD

**Issue by michaelschuster** [graphql-dotnet/server#255]
Hi,

we're using subscriptions and we're seeing an unhandled exception in our log files with the following stack trace:
```
      Connection id "0HLPA441RJJBT", Request id "0HLPA441RJJBT:00000002": An unhandled exception was thrown by the application.
System.Net.WebSockets.WebSocketException (0x80004005): The WebSocket is in an invalid state ('Aborted') for this operation. Valid states are: 'Open, CloseReceived, CloseSent'
   at System.Net.WebSockets.WebSocketValidate.ThrowIfInvalidState(WebSocketState currentState, Boolean isDisposed, WebSocketState[] validStates)
   at System.Net.WebSockets.ManagedWebSocket.CloseAsync(WebSocketCloseStatus closeStatus, String statusDescription, CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at GraphQL.Server.Transports.WebSockets.WebSocketReaderPipeline.Complete(WebSocketCloseStatus closeStatus, String statusDescription)
   at GraphQL.Server.Transports.Subscriptions.Abstractions.SubscriptionServer.Terminate()
   at GraphQL.Server.Transports.WebSockets.WebSocketConnection.Connect()
   at GraphQL.Server.Transports.WebSockets.GraphQLWebSocketsMiddleware`1.InvokeAsync(HttpContext context)
   at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.Invoke(HttpContext context)
   at SimpleInjector.Integration.AspNetCore.RequestScopingStartupFilter.<ConfigureRequestScoping>b__3_0(HttpContext context, Func`1 next)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```

Unfortunately, I was not able to reproduce it locally, I just see the error in the log files. I am not sure if it's related to this issue https://github.com/graphql-dotnet/server/issues/252, 
so I opened a separate ticket.

**Comment by braudabaugh** [comment]
We are seeing this problem too.  Let me know if there is anything I can do to help investigate.

**Comment by jinhong-** [comment]
Hi, for our case, we are experiencing this problem after having it behind load balancers or reverse proxies such as AWS NLB, nginx, etc. They have preconfigured tcp idle timeouts where if the connection is not kept alive, it will be terminated by the load balancer/reverse proxy. Try to time the disconnect and see if it is consistently disconnects at a fixed time when no data is sent between server and client. That was how we found out the cause
The way to extend the lifetime of the connection is to send the WebSockets "ping" message from the client to the server before the configured idle timeout.
It will also be a good idea to cater for situations like unstable internet connectivity, especially for mobile phones on the go where the clients should automatically reconnect when a disconnect happens.
If it is critical for your clients not to miss messages sent during the disconnect, you would want to maintain some sort of subscription token (this will be simpler if you are using backends like kafka, kinesis or nats subscriptions) which will be sent to your client every message as well as sent back to the server during re subscriptions

**Comment by kanonkanin** [comment]
Hi.

Has anyone found a solution to this? We have the same issue (or very similar).
On Localhost its ok, on Azure with Kubernetes and Nginx ingress cntroller, the error appears once every 60 seconds. When client is in use, the error does not appear. I have been looking for Nginx configmap settings we can use, but I have noy found any suitable documentation on the issue. I guess there must be some timeout setting somewhere.

Has anyone found a solution to this? We have the same issue (or very similar).
On Localhost its ok, on Azure with Kubernetes and Nginx ingress controller, the error appears once every 60 seconds. When client is in use, the error does not appear. I have been looking for Nginx configmap settings we can use, but I have not found any suitable documentation on the issue. I guess there must be some timeout setting somewhere.

I also tried the ping trick (once every 500ms)

GraphQL.Server.Transports.WebSockets 3.4.0
GraphQL 2.4.0
GraphQL.Server.Transports.AspNetCore 3.4.0

Here's my stacktrace:
System.Net.WebSockets.WebSocketException (2): The WebSocket is in an invalid state ('Aborted') for this operation. Valid states are: 'Open, CloseReceived, CloseSent'
   at System.Net.WebSockets.WebSocketValidate.ThrowIfInvalidState(WebSocketState currentState, Boolean isDisposed, WebSocketState[] validStates)
   at System.Net.WebSockets.ManagedWebSocket.CloseAsync(WebSocketCloseStatus closeStatus, String statusDescription, CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at GraphQL.Server.Transports.WebSockets.WebSocketReaderPipeline.Complete(WebSocketCloseStatus closeStatus, String statusDescription)
   at GraphQL.Server.Transports.Subscriptions.Abstractions.SubscriptionServer.Terminate()
   at GraphQL.Server.Transports.WebSockets.WebSocketConnection.Connect()
   at GraphQL.Server.Transports.WebSockets.GraphQLWebSocketsMiddleware`1.InvokeAsync(HttpContext context)
   at api.Startup.VersionReponseHeader(HttpContext context, Func`1 next) in /opt/atlassian/pipelines/agent/build/src/api/Startup.cs:line 220
   at Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware.InvokeCore(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)

**Comment by PRGWITHSTYLE** [comment]
Has anyone found a solution to this issue we are on the latest stable build and are seeing this every time we close the client-side of our web application? We are using subscriptions for our communication between our backend and client.

**Comment by jamiewinder** [comment]
I've just encountered this too. In my case, the web socket seems to be getting aborted immediately; literally the first query which'll have opened the web socket is failing with a 400. The server logs tell me:
```
Connection id "0HM3MCJ3CTS5U", Request id "0HM3MCJ3CTS5U:00000002": An unhandled exception was thrown by the application.
System.Net.WebSockets.WebSocketException (0x80004005): The WebSocket is in an invalid state ('Aborted') for this operation. Valid states are: 'Open, CloseReceived, CloseSent'
```

**Comment by DavidKiff** [comment]
We have this too..

It shouldnt throw on the finaliser:
\Transports.Subscriptions.WebSockets\WebSocketReaderPipeline.cs
`
protected ISourceBlock<string> CreateMessageReader()
        {
            IPropagatorBlock<string, string> source = new BufferBlock<string>(
                new ExecutionDataflowBlockOptions
                {
                    EnsureOrdered = true,
                    BoundedCapacity = 1,
                    MaxDegreeOfParallelism = 1
                });

            **Task.Run(async () => { await ReadMessageAsync(source); });**

            return source;
        }
`

The await is doing very little here, since the Task.Run task isn't awaited.  The method doesn't return a Task (ISourceBlock), so we need to update that or block - which obviously wouldn't be an option.

We can mitigate the above exception by not closing something thats aborted (WebSocketReaderPipeline.Complete), but I think there will always be a race without locking, since the check and the close are separate operations.

**Comment by kamnyborg** [comment]
Any news on this? We are having the same issue:

```
2020-12-03T19:48:31.999351041Z: [INFO]  System.Net.WebSockets.WebSocketException (0x80004005): The WebSocket is in an invalid state ('Aborted') for this operation. Valid states are: 'Open, CloseReceived, CloseSent'
2020-12-03T19:48:31.999366341Z: [INFO]     at System.Net.WebSockets.WebSocketValidate.ThrowIfInvalidState(WebSocketState currentState, Boolean isDisposed, WebSocketState[] validStates)
2020-12-03T19:48:32.000351456Z: [INFO]     at System.Net.WebSockets.ManagedWebSocket.CloseAsync(WebSocketCloseStatus closeStatus, String statusDescription, CancellationToken cancellationToken)
2020-12-03T19:48:32.000366756Z: [INFO]  --- End of stack trace from previous location where exception was thrown ---
2020-12-03T19:48:32.000371956Z: [INFO]     at GraphQL.Server.Transports.WebSockets.WebSocketReaderPipeline.Complete(WebSocketCloseStatus closeStatus, String statusDescription)
2020-12-03T19:48:32.000424057Z: [INFO]     at 
GraphQL.Server.Transports.Subscriptions.Abstractions.SubscriptionServer.Terminate()
2020-12-03T19:48:32.000429457Z: [INFO]     at GraphQL.Server.Transports.WebSockets.WebSocketConnection.Connect()
```

Might be a stupid question, but isn't it possible to handle this more gracefully by checking on the websocket state before closing it? We're getting a lot of these on our server throughout the day. This is from one of our QA servers that are used by <10 users (last 24h):
![image](https://user-images.githubusercontent.com/2311623/101082880-7cd65380-35ac-11eb-9eb8-da1e73beb717.png)

**Comment by sungam3r** [comment]
There is no time for this project now. We are preparing GraphQL.NET v4 release due out this week. After that, I plan to deal with the server side.

**Comment by gao-artur** [comment]
We see same problem using Grapiql UI, we don't use subscriptions.

**Comment by Shane32** [comment]
This issue should be fixed in the latest version (5.2.0).  Please re-open if there are further issues.

**Comment by Adarsh7394000** [comment]
ff
